src/Adapter/Presenter/Cart/CartLazyArray.php

```diff
@@ -0,0 +1,604 @@
+                'amount' => $total_discount,
+                'value' => $this->priceFormatter->format($total_discount),
+            ];
+        } else {
+            $subtotals['discounts'] = null;
+        }
+        if ($this->cart->gift) {
+            $giftWrappingPrice = ($this->cart->getGiftWrappingPrice($this->cartPresenter->includeTaxes()) != 0)
+                ? $this->cart->getGiftWrappingPrice($this->cartPresenter->includeTaxes())
+                : 0;
+
+            $subtotals['gift_wrapping'] = [
+                'type' => 'gift_wrapping',
+                'label' => $this->translator->trans('Gift wrapping', [], 'Shop.Theme.Checkout'),
+                'amount' => $giftWrappingPrice,
+                'value' => ($giftWrappingPrice > 0)
+                    ? $this->priceFormatter->convertAndFormat($giftWrappingPrice)
+                    : $this->translator->trans('Free', [], 'Shop.Theme.Checkout'),
+            ];
+        }
+        if (!$this->cart->isVirtualCart()) {
+            $shippingCost = $this->cart->getTotalShippingCost(null, $this->cartPresenter->includeTaxes());
+        } else {
+            $shippingCost = 0;
+        }
+        $subtotals['shipping'] = [
+            'type' => 'shipping',
+            'label' => $this->translator->trans('Shipping', [], 'Shop.Theme.Checkout'),
+            'amount' => $shippingCost,
+            'value' => $this->getShippingDisplayValue($this->cart, $shippingCost),
+        ];
+        $subtotals['tax'] = null;
+        if (Configuration::get('PS_TAX_DISPLAY')) {
+            $total_excluding_tax = $this->cart->getOrderTotal(false);
+            $total_including_tax = $this->cart->getOrderTotal(true);
+            $taxAmount = $total_including_tax - $total_excluding_tax;
+            $subtotals['tax'] = [
+                'type' => 'tax',
+                'label' => ($this->cartPresenter->includeTaxes())
+                    ? $this->translator->trans('Included taxes', [], 'Shop.Theme.Checkout')
+                    : $this->translator->trans('Taxes', [], 'Shop.Theme.Checkout'),
+                'amount' => $taxAmount,
+                'value' => $this->priceFormatter->format($taxAmount),
+            ];
+        }
+
+        $this->subTotals = $subtotals;
+
+        return $this->subTotals;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getProductsCount(): int
+    {
+        // If product list is already available, no need to execute a new sql query
+        if (isset($this->products)) {
+            $this->productsCount = array_reduce($this->products, function ($count, $product) {
+                return $count + $product['quantity'];
+            }, 0);
+
+            return $this->productsCount;
+        }
+
+        // Product list is not available, only query the nb of products
+        $this->productsCount = (int) $this->cart->getNbProducts($this->cart->id);
```

Comment:
We could cache it like this on repeated requests

```suggestion
        if ($this->productsCount === null) {
            // If product list is already available, no need to execute a new sql query
            if (isset($this->products)) {
                $this->productsCount = array_reduce($this->products, function ($count, $product) {
                    return $count + $product['quantity'];
                }, 0);
            // Product list is not available, only query the nb of products
            } else {
                $this->productsCount = (int) $this->cart->getNbProducts($this->cart->id);
            }
        }
```

Reply:
same as below: these getters are not to be called directly, the presenter is meant to be used as an array, in templates, or having its properties called directly

the AbstractLazyArray is responsible for caching the result and calling the getter only if the data is not cached already